File: cpp/cluster.cpp


#include "cluster.h"
#include <algorithm>
#include <cmath>
#include <vector>
#include <list>
#include <iostream>

using namespace std;

double Cluster::getCost() {
    return cost;
}

Cluster::Cluster(int d, double t, int n):
    points(list< Point*>()), cost(0), dim_counts(new int[d]),
    point_count(0), cost_parts(new double[d]), representative(Point()),
    rep_sum(0), dis_sum(0), dim(d), dim_buckets(new set<int>[n]), t(t){
    for (int i=0;i<d;++i) {
        cost_parts[i] = 0.0;
        dim_counts[i] = 0;
    }
}

//utility function
static double nlogn(int n) {
    return n==0 ? 0 : log(n)*n;
}

void Cluster::addPoint( Point* point) {
    getCostWith(point, true);
    points.push_back(point);
    for (Point::iterator it = point->begin(); it != point->end(); ++it) {
        if (dim_counts[*it] != 0) {
            dim_buckets[dim_counts[*it]].erase(*it);
        }
        ++dim_counts[*it];
        dim_buckets[dim_counts[*it]].insert(*it);
    }
    ++point_count;
}

void Cluster::removePoint( Point* point) {
    getCostWithout(point, true);
    points.remove(point);
    for (Point::iterator it = point->begin(); it != point->end(); ++it) {
        dim_buckets[dim_counts[*it]].erase(*it);
        --dim_counts[*it];
        if (dim_counts[*it] != 0) {
            dim_buckets[dim_counts[*it]].insert(*it);
        }
    }
    --point_count;
}

//utility function, removes the first occurence of a value in a vector
static void removeFromVector(Point& point, int val) {
    for (Point::iterator it = point.begin(); it != point.end(); ++it) {
        if (*it == val) {
            point.erase(it);
            break;
        }
    }
}

double Cluster::getCostWith( Point* point, bool update) {
    double cost_ = this->cost;              //local copies of the cluster variables
    int dis_sum_ = this->dis_sum;
    int rep_sum_ = this->rep_sum;
    vector<int> difference_set(representative.size());

    vector<int>::iterator itend = set_difference(representative.begin(),
        representative.end(), point->begin(), point->end(), difference_set.begin());
    for (vector<int>::iterator it = difference_set.begin(); it != itend; ++it) {
        //position is in the representative but not in the point being added
        if (dim_counts[*it] < t*(point_count+1)) {
            //new point causes the position to be removed from the representative
            dis_sum_ += 2*dim_counts[*it]-point_count;
            cost_ -= nlogn(dim_counts[*it]) - cost_parts[*it];
            --rep_sum_;
            if (update) {
                removeFromVector(representative, *it);
                cost_parts[*it] = nlogn(dim_counts[*it]);
            }
        } else {
            //new point does not cause a change in the representative
            ++dis_sum_;
            cost_ -= nlogn(point_count - dim_counts[*it]+1) - cost_parts[*it];
            if (update)
                cost_parts[*it] = nlogn(point_count - dim_counts[*it]+1);
        }
    }

    for (Point::iterator it = point->begin(); it != point->end(); ++it) {
        if (dim_counts[*it]+1 < t*(point_count+1)) {
            //the position is not in the representative but is in the new point
            ++dis_sum_;
            cost_ -= nlogn(dim_counts[*it]+1) - cost_parts[*it];
            if (update)
                cost_parts[*it] = nlogn(dim_counts[*it]+1);
        } else if (dim_counts[*it] < t*point_count || point_count==0) {
            //the position is in the new point, which causes it
            //to cross the treshold and be added to the representative
            dis_sum_ += point_count-2*dim_counts[*it];
            cost_ -= nlogn(point_count-dim_counts[*it]) - cost_parts[*it];
            ++rep_sum_;
            if (update) {
                representative.push_back(*it);
                cost_parts[*it] = nlogn(point_count-dim_counts[*it]);
            }
        }
    }

    cost_ += nlogn(rep_sum_) - nlogn(rep_sum);
    cost_ += nlogn(dis_sum_) - nlogn(dis_sum);
    if (update) {
        cost = cost_;
        rep_sum = rep_sum_;
        dis_sum = dis_sum_;
    }
    return cost_;
}

//utility function, returns whether a value is present in a vector
static bool valInPoint( Point* point, int val) {
    for (Point::iterator it = point->begin(); it != point->end(); ++it) {
        if (*it == val) {
            return true;
        }
    }
    return false;
}

double Cluster::getCostWithout( Point* point, bool update) {
    double cost_ = this->cost;              //local copies of the cluster variables
    int dis_sum_ = this->dis_sum;
    int rep_sum_ = this->rep_sum;
    vector<int> difference_set(representative.size());

    vector<int>::iterator itend = set_difference(representative.begin(),
        representative.end(), point->begin(), point->end(), difference_set.begin());
    for (vector<int>::iterator it = difference_set.begin(); it != itend; ++it) {
        //position is in the representative but not in the point being removed
        --dis_sum_;
        cost_ -= nlogn(point_count - dim_counts[*it] - 1) - cost_parts[*it];
        if (update)
               cost_parts[*it] = nlogn(point_count - dim_counts[*it] - 1);
    }

    for (Point::iterator it = point->begin(); it != point->end(); ++it) {
        if (dim_counts[*it] < t*point_count) {
            //position is in the removed point but does not change the representative
            --dis_sum_;
            cost_ -= nlogn(dim_counts[*it]-1) - cost_parts[*it];
            if (update)
                cost_parts[*it] = nlogn(dim_counts[*it]-1);
        } else if (dim_counts[*it]-1 < t*(point_count-1)) {
            //position in the removed point causes it to be removed from the representative
            dis_sum_ += 2*dim_counts[*it] - point_count - 1;
            cost_ -= nlogn(dim_counts[*it] - 1) - cost_parts[*it];
            --rep_sum_;
            if (update) {
                removeFromVector(representative, *it);
                cost_parts[*it] = nlogn(dim_counts[*it] - 1);
            }
        }
    }

    //special case: checking if removing the point would
    //cause a position to be added to the representative
    int border_val = (int)(point_count*t-0.0000001);
    if (border_val >= (point_count-1)*t) {
        set<int>& bucket = dim_buckets[border_val];
        for (set<int>::iterator it = bucket.begin(); it != bucket.end(); ++it)
            if (valInPoint(point, *it)) {
                dis_sum_ += point_count-2*dim_counts[*it]-1;
                ++rep_sum_;
                cost_ -= nlogn(point_count-dim_counts[*it]-1) - cost_parts[*it];
                if (update) {
                    representative.push_back(*it);
                    cost_parts[*it] = nlogn(point_count-dim_counts[*it]-1);
                }
            }
    }

    cost_ += nlogn(rep_sum_) - nlogn(rep_sum);
    cost_ += nlogn(dis_sum_) - nlogn(dis_sum);
    if (update) {
        cost = cost_;
        rep_sum = rep_sum_;
        dis_sum = dis_sum_;
    }
    return cost_;
}

File: python/Cluster.py
from math import log

def nlogn(n):
    return n*log(n) if n!=0 else 0

def add_count(count, diction, pos):
    if count in diction:
        diction[count].add(pos)
    else:
        diction[count] = set([pos])

class Cluster(object):
    def __init__(self, dim, i, t):
        self.points = []                                #list of points in the cluster
        self.cost = 0.0                                 #the cost of the cluster
        self.dim_counts = [0 for _ in range(dim)]       #the number of 1s in each column
        self.point_count = 0                            #the total number of points
        self.cost_parts = [0.0 for _ in range(dim)]     #the saved parts of the cost that pertain to a certain column
        self.representative = set()                     #the point that represents the cluster
        self.rep_sum = 0                                #the number of 1s in the representative
        self.dis_sum = 0                                #the sum of the differences between the representative and the cluster points
        self.dim = dim                                  #the dimension of the space
        self.number = i
        self.dim_buckets = dict()
        self.t = t

    def __str__(self):                                  #string presentation for debugging purposes
        return str(self.number) + ":" + str(self.point_count) + " " + str(self.dim_counts) + \
            " " + str(self.representative) + " " + str(self.dis_sum) + "\n" + str(self.points) + "\n"

    def addPoint(self, point):                          #adds a point
        self.getCostWith(point, update=True)
        self.points.append(point)
        for pos in point:
            if self.dim_counts[pos] != 0:
                self.dim_buckets[self.dim_counts[pos]].remove(pos)
            self.dim_counts[pos] += 1
            add_count(self.dim_counts[pos], self.dim_buckets, pos)
        self.point_count += 1

    def removePoint(self, point):                       #removes a point
        self.getCostWithout(point, update=True)
        self.points.remove(point)
        for pos in point:
            self.dim_buckets[self.dim_counts[pos]].remove(pos)
            self.dim_counts[pos] -= 1
            add_count(self.dim_counts[pos], self.dim_buckets, pos)
        self.point_count -= 1

    def getCostWith(self, point, update=False):
        """
        Returns the cost that would occur if the point was added
        if update=True, it updates the internal values with the modifications
        observed by the adding of the point
        """
        cost = self.cost
        dis_sum = self.dis_sum
        rep_sum = self.rep_sum
        difference_set = self.representative - point
        for pos in difference_set:
            #position is in the representative but not in the point being added
            if self.dim_counts[pos] <= self.t*(self.point_count+1):
                #new point causes the position to be removed from the representative
                dis_sum += 2*self.dim_counts[pos]-self.point_count
                cost -= nlogn(self.dim_counts[pos]) - self.cost_parts[pos]
                rep_sum -= 1
                if update:
                    self.representative.remove(pos)
                    self.cost_parts[pos] = nlogn(self.dim_counts[pos])
            else:
                #new point does not cause a change in the representative
                dis_sum += 1
                cost -= nlogn(self.point_count - self.dim_counts[pos]+1) - self.cost_parts[pos]
                if update:
                    self.cost_parts[pos] = nlogn(self.point_count - self.dim_counts[pos]+1)
        for pos in point:
            if self.dim_counts[pos]+1 <= self.t*(self.point_count+1):
                #the position is not in the representative but is in the new point
                dis_sum += 1
                cost -= nlogn(self.dim_counts[pos]+1) - self.cost_parts[pos]
                if update:
                    self.cost_parts[pos] = nlogn(self.dim_counts[pos]+1)
            elif self.dim_counts[pos] <= self.t*self.point_count or self.point_count==0:
                #the position is in the new point, which causes it
                #to cross the treshold and be added to the representative
                dis_sum += self.point_count-2*self.dim_counts[pos]
                cost -= nlogn(self.point_count-self.dim_counts[pos]) - self.cost_parts[pos]
                rep_sum += 1
                if update:
                    self.representative.add(pos)
                    self.cost_parts[pos] = nlogn(self.point_count-self.dim_counts[pos])
        cost += nlogn(rep_sum) - nlogn(self.rep_sum)
        #cost += (rep_sum - self.rep_sum)*log(self.dim)
        cost += nlogn(dis_sum) - nlogn(self.dis_sum)
        if update:
            self.cost = cost
            self.rep_sum = rep_sum
            self.dis_sum = dis_sum
        return cost

    def getCostWithout(self, point, update=False):
        """
        Returns the cost that would occur if the point was removed
        if update=True, it updates the internal values with the modifications
        observed by the removal of the point
        """
        cost = self.cost
        dis_sum = self.dis_sum
        rep_sum = self.rep_sum
        for pos in self.representative - point:
            #position is in the representative but not in the point being removed
            dis_sum -= 1
            cost -= nlogn(self.point_count - self.dim_counts[pos] - 1) - self.cost_parts[pos]
            if update:
                self.cost_parts[pos] = nlogn(self.point_count - self.dim_counts[pos] - 1)
        for pos in point:
            if self.dim_counts[pos] <= self.t*self.point_count:
                #position is in the removed point but does not change the representative
                dis_sum -= 1
                cost -= nlogn(self.dim_counts[pos]-1) - self.cost_parts[pos]
                if update:
                    self.cost_parts[pos] = nlogn(self.dim_counts[pos]-1)
            elif self.dim_counts[pos]-1 <= self.t*(self.point_count-1):
                #position in the removed point causes it to be removed from the representative
                dis_sum += 2*self.dim_counts[pos]-self.point_count-1
                cost -= nlogn(self.dim_counts[pos]-1) - self.cost_parts[pos]
                rep_sum -= 1
                if update:
                    self.representative.remove(pos)
                    self.cost_parts[pos] = nlogn(self.dim_counts[pos]-1)
        #special case: checking if removing the point would
        #scause a position to be added to the representative
        border_val = int(self.point_count*self.t-0.0000001)
        if border_val >= (self.point_count-1)*self.t:
            for i in self.dim_buckets.get(border_val, set()):
                if i not in point:
                    dis_sum += self.point_count-2*self.dim_counts[i]-1
                    rep_sum += 1
                    cost -= nlogn(self.point_count-self.dim_counts[i]-1) - self.cost_parts[i]
                    if update:
                        self.representative.add(i)
                        self.cost_parts[i] = nlogn(self.point_count-self.dim_counts[i]-1)
        cost += nlogn(rep_sum) - nlogn(self.rep_sum)
        #cost += (rep_sum - self.rep_sum)*log(self.dim)
        cost += nlogn(dis_sum) - nlogn(self.dis_sum)
        if update:
            self.cost = cost
            self.dis_sum = dis_sum
            self.rep_sum = rep_sum
        return cost

    def verify(self):
        """
        debug function, checks the integrity of the calculated values, used for debug purposes
        """
        goods = []
        if not all(goods):
            blad
        goods.append(self.point_count == len(self.points))
        if not all(goods):
            blad
        goods.append(sum(len(point) for point in self.points) == sum(val for val in self.dim_counts))
        if not all(goods):
            blad
        goods.append(len(self.representative) == self.rep_sum)
        if not all(goods):
            blad
        goods.append(all(i in self.representative for i, count in enumerate(self.dim_counts) if count >= self.t*self.point_count))
        if not all(goods):
            blad
        goods.append(all(i not in self.representative for i, count in enumerate(self.dim_counts) if count < self.t*self.point_count))
        if not all(goods):
            blad
        goods.append(self.dis_sum == sum((self.point_count - count) if i in self.representative else count for i, count in enumerate(self.dim_counts)))
        if not all(goods):
            print self, sum((self.point_count - count) if i in self.representative else count for i, count in enumerate(self.dim_counts))
            blad
